51-android.rules

```diff
@@ -179,87 +179,86 @@ ATTR{idVendor}!="18d1", GOTO="not_Google"
+ATTR{idProduct}=="5203", GOTO="adb"
+ATTR{idProduct}=="5208", GOTO="adb"
+
+ATTR{idProduct}=="2d00", GOTO="adb"
+ATTR{idProduct}=="2d01", GOTO="adb"
+ATTR{idProduct}=="2d03", GOTO="adb"
+ATTR{idProduct}=="2d05", GOTO="adb"
 #   Nexus 7
-ATTR{idProduct}=="4e42", ENV{adb_adb}="yes"
-ATTR{idProduct}=="4e40", ENV{adb_adbfast}="yes"
+ATTR{idProduct}=="4e42", GOTO="adb"
+ATTR{idProduct}=="4e40", GOTO="adbfast"
 #   Nexus 5, Nexus 10
-ATTR{idProduct}=="4ee1", ENV{adb_adbfast}="yes"
+ATTR{idProduct}=="4ee1", GOTO="adbfast"
 #   Nexus S (4e22=mass_storage,adb 4e24=rndis,adb)
 #   See https://android.googlesource.com/device/samsung/crespo/+/android-4.1.2_r2.1/init.herring.usb.rc
-ATTR{idProduct}=="4e22", ENV{adb_adb}="yes"
-ATTR{idProduct}=="4e24", ENV{adb_adb}="yes"
-ATTR{idProduct}=="4e20", ENV{adb_adbfast}="yes"
+ATTR{idProduct}=="4e22", GOTO="adb"
+ATTR{idProduct}=="4e24", GOTO="adb"
+ATTR{idProduct}=="4e20", GOTO="adbfast"
 #   Galaxy Nexus, Galaxy Nexus (GSM)
-ATTR{idProduct}=="4e30", ENV{adb_adbfast}="yes"
+ATTR{idProduct}=="4e30", GOTO="adbfast"
 #   Nexus One (4e11=normal,4e12=debug,0fff=debug)
-ATTR{idProduct}=="4e12", ENV{adb_adb}="yes"
-ATTR{idProduct}=="0fff", ENV{adb_adbfast}="yes"
+ATTR{idProduct}=="4e12", GOTO="adb"
+ATTR{idProduct}=="0fff", GOTO="adbfast"
 #   Generic and unspecified debug interface (test after d00?)
 #   examples: Xiaomi Mi/Redmi 2, Anbernic RG353P
-ATTR{idProduct}=="d00d", ENV{adb_adbfast}="yes"
+ATTR{idProduct}=="d00d", GOTO="adbfast"
 #   Recovery adb entry for Nexus Family (orig d001, OP3 has 18d1:d002)
-ATTR{idProduct}=="d00?", ENV{adb_adb}="yes"
+ATTR{idProduct}=="d00?", GOTO="adb"
 
 # Other vendors that also used duplicated Google's idVendor code follows:
 # IDEA XDS-1078 (debug=2c11)
-ATTR{idProduct}=="2c11", ENV{adb_adb}="yes"
-GOTO="android_usb_rule_match"
+ATTR{idProduct}=="2c11", GOTO="adb"
+GOTO="android_usb_rules_end"
 LABEL="not_Google"
 
 # Haier
-ATTR{idVendor}=="201e", ENV{adb_user}="yes"
+ATTR{idVendor}=="201e", GOTO="user"
 
 # Hisense (includes Fairphone 1)
-ATTR{idVendor}=="109b", ENV{adb_user}="yes"
+ATTR{idVendor}=="109b", GOTO="user"
 
 # Honeywell/Foxconn
 ATTR{idVendor}!="0c2e", GOTO="not_Honeywell"
 #   D70e
-ATTR{idProduct}=="0ba3", ENV{adb_adb}="yes"
-GOTO="android_usb_rule_match"
+ATTR{idProduct}=="0ba3", GOTO="adb"
+GOTO="android_usb_rules_end"
 LABEL="not_Honeywell"
 
 # HTC
 ATTR{idVendor}!="0bb4", GOTO="not_HTC"
-ENV{adb_user}="yes"
 #   fastboot mode enabled
-ATTR{idProduct}=="0fff", ENV{adb_adbfast}="yes", GOTO="android_usb_rule_match"
+ATTR{idProduct}=="0fff", GOTO="adbfast"
 #   ADP1, Dream, G1, HD2, Magic, Tatoo (0c01=mass_storage)
 #   NOTE: Amazon Kindle 8 2016 (giza) (fastboot=0bb4:0c01 conflicts with mass storage=0c01)
-ATTR{idProduct}=="0c02", ENV{adb_adbfast}="yes"
+ATTR{idProduct}=="0c02", GOTO="adbfast"
 #   ChaCha
-ATTR{idProduct}=="0cb2", ENV{adb_adbfast}="yes"
+ATTR{idProduct}=="0cb2", GOTO="adbfast"
 #   Desire (Bravo)
 ATTR{idProduct}=="0c87", SYMLINK+="android_adb"
```

Comment:
The ENV{adb_adbfast} and ENV{adb_adb} are easy to see as GOTO="adbfast" and GOTO="adb".

SYMLINK+="android_adb" might be a GOTO something with ADB, or maybe not with ADB. Hard to guess.
Not enough info at the moment to know just by looking, unless you find other info elsewhere.
To keep this rather big patch straightforward, it might be best to leave it as-is just for this patch, and then make that other mod afterwards?

Reply:
Thanks for the explanation. I guess we would need to dig through old commits and issues. It still likely won't resolve all the doubts but may cut down the number of these unclear cases.

Agreed that it's probably better to do that in some later commits.

Reply:
Agreed. More info always helps.
I noticed several (adb+mtp) in 69-libmtp.rules while making PR 171 in libmtp/libmtp, sometimes looking elsewhere for more info might help, or not.